### PR TITLE
feat: add support for shutter proposals

### PR DESCRIPTION
**Pull request by wa0x6e**
### Summary

Add support for shutter proposals

Closes: #148 

#### For active shutter proposals

1. Go to an active shutter proposal (e.g. http://localhost:8080/#/s:test.wa0x6e.eth/proposal/0x9623af39201332732f38227e75deced57fa4cc53f6bbb5380d0bdce40172d0e0)
2. It should not show the results in the sidebar, and show the Shutter block instead

Screenshot 2024-03-14 at 11 35 06

In the votes pages, it should show an "Encrypted choice" message instead of the choice

Screenshot 2024-03-14 at 11 35 00

In the votes modal, it should also hide the choice behind an "Encryped choice" text (and hide the background progress bar)

Screenshot 2024-03-15 at 21 55 08

#### Voting

1. Cast a vote on a shutter proposal
2. It should cast the vote (confirm on the dev tools network tab, call to seq.snapshot.org should return a 200 code)

#### For closed shutter proposals

1. Go to a closed shutter proposals (e.g. http://localhost:8080/#/s:test.wa0x6e.eth/proposal/0x5ddbe81f1870a5154bd7b120bdc2c1755315157e7b230d38c74544ddd1ac62ef)
4. It should show the results as usual, as well as the "Powered by Shutter footer"

Screenshot 2024-03-14 at 11 35 20

### To-Do

- [x] Fixed tooltip overflow hidden when inside result (see https://github.com/snapshot-labs/sx-monorepo/pull/170)
- [x] Enable shutter voting from UI
- [x] Handle missing typescript definition for shutter-network package more elegantly
- [x] Add tests for shutter voting (need to vote on mainnet instead of testnet)

### CODE REVIEW

**Comment by wa0x6e** [comment]
There's one bug left to fix, related to tooltip clipping, but not really related to this PR

**Comment by samuveth** [comment]
1. Choice is missing on votes page, should show lock icon and maybe some text like "Encrypted choice"
2. Should probably also show some text on `Results` section like "Encrypted"

**Comment by wa0x6e** [comment]
@Sekhmet Any ideas on how to handle the non-existing declaration file for the new shutter dependency ? It does not ship with typescript support, and does not exists on DefinitelyTyped

File: apps/ui/src/components/Encrypted.vue [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    iconOnly?: boolean;
+  }>(),
+  { encrypted: false, iconOnly: false }
+);
+</script>
+
+<template>
+  <div
+    v-if="encrypted"
+    v-tippy="{
```
Comment (samuveth) [inline apps/ui/src/components/Encrypted.vue:14]: We should use `v-tippy` inside `UiTooltip` and use `UiTooltip` here instead I think

File: apps/ui/src/assets/css/tippy.scss [hunk 2]
```diff
@@ -0,0 +1,87 @@
+// V-Tippy tooltip package styling - could be moved somewhere else?
```
Comment (samuveth) [inline apps/ui/src/assets/css/tippy.scss:1]:
```suggestion
```

File: apps/ui/src/views/Proposal/Votes.vue [hunk 3]
```diff
@@ -209,26 +210,28 @@ watch([sortBy, choiceFilter], () => {
...
-                  '!text-skin-success !border-skin-success': vote.choice === 1,
-                  '!text-skin-danger !border-skin-danger': vote.choice === 2,
-                  '!text-gray-500 !border-gray-500': vote.choice === 3
-                }"
-              >
-                <IH-check v-if="vote.choice === 1" class="inline-block" />
-                <IH-x v-else-if="vote.choice === 2" class="inline-block" />
-                <IH-minus-sm v-else class="inline-block" />
-              </UiButton>
+              <Encrypted :encrypted="isEncrypted">
```
Comment (samuveth) [inline apps/ui/src/views/Proposal/Votes.vue:213]: What I don't really understand is why you need to wrap this inside `Encrypted` can't you simply use `v-if=isEncrypted` and `v-else`?

File: apps/ui/src/components/Encrypted.vue (hunk 1 above)
Comment (wa0x6e) [reply apps/ui/src/components/Encrypted.vue:14]: Will move everything back as UiTooltip, and do the v-tippy migration in a separate PR

**Comment by Sekhmet** [comment]
> @Sekhmet Any ideas on how to handle the non-existing declaration file for the new shutter dependency ? It does not ship with typescript support, and does not exists on DefinitelyTyped

We could ship our own as those are fairly simple.

```diff
diff --git a/packages/sx.js/src/clients/offchain/utils.ts b/packages/sx.js/src/clients/offchain/utils.ts
index 39d82c3f..798185b9 100644
--- a/packages/sx.js/src/clients/offchain/utils.ts
+++ b/packages/sx.js/src/clients/offchain/utils.ts
@@ -2,9 +2,7 @@ import { randomBytes } from '@ethersproject/random';
 import { BigNumber } from '@ethersproject/bignumber';
 import { arrayify, hexlify } from '@ethersproject/bytes';
 import { toUtf8Bytes, formatBytes32String } from '@ethersproject/strings';
-// @ts-ignore
 import shutterWasm from '@shutter-network/shutter-crypto/dist/shutter-crypto.wasm?url';
-// @ts-ignore
 import { init, encrypt } from '@shutter-network/shutter-crypto';
 import type { Choice } from './types';
 import type { Privacy } from '../../types';
diff --git a/packages/sx.js/src/custom.d.ts b/packages/sx.js/src/custom.d.ts
new file mode 100644
index 00000000..c49a6fa1
--- /dev/null
+++ b/packages/sx.js/src/custom.d.ts
@@ -0,0 +1,14 @@
+declare module '@shutter-network/shutter-crypto/dist/shutter-crypto.wasm?url' {
+  const wasm: any;
+  export default wasm;
+}
+
+declare module '@shutter-network/shutter-crypto' {
+  export function init(wasm: string): Promise<void>;
+  export function encrypt(
+    message: Uint8Array,
+    eonPublicKey: Uint8Array,
+    epochId: Uint8Array,
+    sigma: Uint8Array
+  ): Promise<Uint8Array>;
+}
```

**Comment by wa0x6e** [comment]
Getting back to draft in order to implement the new UI

**Comment by wa0x6e** [comment]
PR ready again, implementing the new UI from designer. Please read the first post again for updated description.

File: apps/ui/src/components/ProposalResults.vue [hunk 4]
```diff
@@ -45,83 +45,127 @@ const results = computed(() =>
...
+    </div>
+
+    <div class="flex flex-col gap-1">
+      <div class="text-skin-heading font-semibold">Shutter privacy enabled</div>
+      All votes will be encrypted until the voting period has ended and the final score is
+      calculated.
+    </div>
+
+    <div class="mt-2.5 text-center">
+      <a
```
Comment (samuveth) [inline apps/ui/src/components/ProposalResults.vue:65]: Should open in a new tab I think

File: apps/ui/src/components/ProposalResults.vue [hunk 5]
```diff
@@ -45,83 +45,127 @@ const results = computed(() =>
 </script>
 
 <template>
-  <div v-if="proposal.type !== 'basic'">
+  <div
+    v-if="!!props.proposal.privacy && !props.proposal.completed"
+    class="text-center p-3.5 leading-5"
```
Comment (samuveth) [inline apps/ui/src/components/ProposalResults.vue:50]: There is already padding on parent, no need for additional padding on x-axis here.

File: apps/ui/src/views/Proposal/Votes.vue [hunk 6]
```diff
@@ -209,26 +209,32 @@ watch([sortBy, choiceFilter], () => {
...
             <td class="relative">
-              <div
-                v-if="proposal.type !== 'basic'"
-                class="truncate"
-                :title="proposal.choices[vote.choice - 1]"
-              >
-                {{ proposal.choices[vote.choice - 1] }}
+              <div v-if="!!props.proposal.privacy && !props.proposal.completed">
+                <span class="text-skin-heading">Encrypted choice</span>
+                <i-h-lock-closed class="inline-block ml-1" />
```
Comment (samuveth) [inline apps/ui/src/views/Proposal/Votes.vue:214]: Icon doesn't seem properly aligned 
image

Comment (samuveth) [reply apps/ui/src/views/Proposal/Votes.vue:214]: Add seems bigger than in the design

**Comment by samuveth** [comment]
Icon still not properly centered for me 
Screenshot 2024-03-14 at 4 24 35 in the afternoon
Screenshot 2024-03-14 at 4 24 41 in the afternoon

**Review by Sekhmet — APPROVED** [review]
tACK

We can adjust UI later if we have finished designs.

**Comment by Sekhmet** [comment]
@wa0x6e could you resolve conflicts and merge this PR?

File: apps/ui/src/views/Proposal/Votes.vue [hunk 7]
```diff
@@ -209,26 +209,39 @@ watch([sortBy, choiceFilter], () => {
...
-                {{ proposal.choices[vote.choice - 1] }}
+              <template v-if="!!props.proposal.privacy && !props.proposal.completed">
+                <div class="hidden md:block">
+                  <div class="flex gap-1 items-center">
+                    <span class="text-skin-heading">Encrypted choice</span>
+                    <i-h-lock-closed class="w-[16px] h-[16px] shrink-0" />
+                  </div>
+                </div>
+                <UiTooltip title="Encrypted choice" class="cursor-help md:hidden">
+                  <i-h-lock-closed class="w-[16px] h-[16px]" />
```
Comment (Sekhmet) [inline apps/ui/src/views/Proposal/Votes.vue:220]:
```suggestion
                    <IH-lock-closed class="w-[16px] h-[16px] shrink-0" />
                  </div>
                </div>
                <UiTooltip title="Encrypted choice" class="cursor-help md:hidden">
                  <IH-lock-closed class="w-[16px] h-[16px]" />
```

This is convention we follow in this repo.

**Comment by wa0x6e** [comment]
Still missing an update on the vote modal, wip

**Comment by wa0x6e** [comment]
> Still missing an update on the vote modal, wip

Fixed the missing update in the votes modal, should looks like this now

Screenshot 2024-03-15 at 21 55 08

File: apps/ui/src/views/Proposal/Votes.vue (hunk 7 above)
Comment (samuveth) [reply apps/ui/src/views/Proposal/Votes.vue]: Which is mixing two different conventions btw :D

Comment (samuveth) [reply apps/ui/src/views/Proposal/Votes.vue]: Ideally we should be using `CamelCase` everywhere, even for things like `router-link` https://vuejs.org/style-guide/rules-strongly-recommended.html#self-closing-components

Comment (wa0x6e) [reply apps/ui/src/views/Proposal/Votes.vue]: is there an eslint rule to enforce this ?

Comment (samuveth) [reply apps/ui/src/views/Proposal/Votes.vue]: I mean `PascalCase`

Comment (samuveth) [reply apps/ui/src/views/Proposal/Votes.vue]: > is there an eslint rule to enforce this ?

Yes, here https://eslint.vuejs.org/rules/component-name-in-template-casing.html

**Comment by samuveth** [comment]
What is the votes modal used for? I don't see a way to open it.

**Comment by wa0x6e** [comment]
> What is the votes modal used for? I don't see a way to open it.

Had a hard time finding it to, go to a proposal and click on "x votes" in the footer

**Comment by wa0x6e** [comment]
> > What is the votes modal used for? I don't see a way to open it.
> 
> Had a hard time finding it too, go to a proposal and click on "x votes" in the footer

Because all the other proposals footer with "x votes" don't trigger this modal

**Review by Sekhmet — APPROVED** [review]
tACK

**Pull request merged**